--- binutils-2.21.51.0.8/ld/ldmain.c.defaults~	2011-04-08 20:58:19.000000000 +0200
+++ binutils-2.21.51.0.8/ld/ldmain.c	2011-05-04 22:13:04.772677095 +0200
@@ -274,9 +274,14 @@ main (int argc, char **argv)
 
   link_info.allow_undefined_version = TRUE;
   link_info.keep_memory = TRUE;
+  link_info.optimize = TRUE;
+  link_info.new_dtags = TRUE;
   link_info.combreloc = TRUE;
+  link_info.relro = TRUE;
   link_info.strip_discarded = TRUE;
-  link_info.emit_hash = TRUE;
+  link_info.emit_hash = FALSE;
+  link_info.emit_gnu_hash = TRUE;
+  link_info.emit_note_gnu_build_id = xstrdup("sha1");
   link_info.callbacks = &link_callbacks;
   link_info.input_bfds_tail = &link_info.input_bfds;
   /* SVR4 linkers seem to set DT_INIT and DT_FINI based on magic _init
